### PR TITLE
feature: supports the OceanBase Oracle kernel in AT mode

### LINKED ISSUE #4620: AT mode supports the OceanBase Oracle kernel
Description: 
Seata AT transaction mode supports MySQL, PostgreSQL, Oracle, TiDB, MariaDB, etc.

Task: 
Based on the existing DB adaptation in AT mode, compile the structure diagram to implement oracle kernel adaptation of Oceanbase database, including sqlParser and driver adaptation, and verify database adaptation using sample.

Skills required: 
Familiar with Java

Expected result:
Oracle kernel adaptation for the Oceanbase database and related verification sample are complete.

---

描述：目前Seata AT模式支持MySQL、PostgreSQL、Oracle、TiDB、MariaDB等，需要对Oceanbase数据库的oracle内核进行相关的支持。

任务: 参照之前的AT模式中DB适配，整理出结构图，实现Oceanbase 数据库的oracle内核适配，主要包含通用的sqlparser和driver适配，并能通过sample对内核适配进行验证。

技能要求：熟悉Java

预期结果：完成Oceanbase 数据库的oracle内核适配和相关验证sample。

### PR DESCRIPTION
**Pull request by hsien999**
- [ ] I have registered the PR [changes](https://github.com/seata/seata/tree/develop/changes).

### Ⅰ. Describe what this PR did
[Summer code 2022] AT mode supports the OceanBase Oracle kernel
1. support sql parser (module sqlparser)
2. support driver adaptation, executors and undo log executors etc (module rm-datasource)
3. support lock & log store in server site (module core)

### Ⅱ. Does this pull request fix one issue?

fixes #4620 

### Ⅳ. Describe how to verify it
A simple project for integration test: [seata-oceanbase-demo](https://github.com/hsien999/seata-oceanbase-demo)

### CODE REVIEW

**Review by a364176773 — COMMENTED** [review]
oceanbase本身高度兼容了oracle和100%兼容5.7的mysql，建议直接继承oracle和mysql的实现，避免出现一套逻辑N处维护的成本，差异点出现了后再做重写也来得及，包括server端的store也一样

File: rm-datasource/src/main/java/io/seata/rm/datasource/util/JdbcUtils.java [hunk 1]
```diff
@@ -47,44 +54,76 @@ static DbTypeParser getDbTypeParser() {
...
+            String databaseName = conn.getMetaData().getDatabaseProductName().toLowerCase();
+            if (JdbcConstants.MYSQL.contains(databaseName)) {
+                return JdbcConstants.MYSQL;
+            } else if (JdbcConstants.ORACLE.contains(databaseName)) {
+                return JdbcConstants.OCEANBASE_ORACLE;
+            } else {
+                try (Statement statement = conn.createStatement();
+                     ResultSet rs = statement.executeQuery("SELECT * FROM DUAL")) {
+                    if (!rs.next()) {
+                        throw new SQLException("Validation query for OceanBase(Oracle mode) didn't return a row");
```
Comment (a364176773) [inline rm-datasource/src/main/java/io/seata/rm/datasource/util/JdbcUtils.java:95]: 这个异常抛出去又被catch没啥一样呀，要不就打印成日志，if 如果没有就改为mysql如果有就是oracle

**Comment by jiayaoO3O** [comment]
请问当前的seata对于oceanbase的支持进度到哪里了? 打算哪个版本支持该功能

**Comment by dargoner** [comment]
> 请问当前的seata对于oceanbase的支持进度到哪里了? 打算哪个版本支持该功能

大家都比较期待

**Comment by funky-eyes** [comment]
Due to code conflicts and age, if you are still interested in participating in the community, please resubmit this pr to the 2.x branch after resolving the conflict.
由于代码冲突以及年代久远，如果你还有兴趣参与到社区，请解决完冲突后重新请提交这份pr至2.x分支中

**Pull request closed without merging**